### PR TITLE
Removing Fake-Layers in UDS

### PR DESCRIPTION
**Pull request by polybassa**
@gpotter2 @guedou 
This PR is a PoC to remove the fake layers in automotive protocols without losing the advantages.

Here are some examples:

Before:
```python
pkt = UDS()/UDS_RDBI(identifiers=[3])
print(repr(pkt))
>>> <UDS  service=ReadDataByIdentifier |<UDS_RDBI  identifiers=[0x3] |>>
```

Now:
```python
pkt = UdsRdbi(identifiers=[3])
print(repr(pkt))
>>> <UdsRdbi  identifiers=[0x3] |>
```

Before:
```python
pkt = UDS()/UDS_RDBIPR()/Payload3(ip="127.0.0.1")
print(repr(pkt))
>>> <UDS  service=ReadDataByIdentifierPositiveResponse |<UDS_RDBIPR  |<Payload3  ip=127.0.0.1 |>>>
```

Now:
```python
class Payload3(Packet):
    fields_desc = [
        IPField("ip", 0),
        IPField("mask", 1)
    ]

UdsRdbiPr.bind_data_record(Payload3, 0x3)

pkt = UdsRdbiPr(dataRecord=Payload3(ip="127.0.0.1")
print(repr(pkt))
>>> <UdsRdbiPr  dataIdentifier=0x3 dataRecord=<Payload3  ip=127.0.0.1 |> |>
```

Before:
```python
>>> pkt = UDS(bytes.fromhex("620003c0a8b201fffffc00"))
>>> print(repr(pkt))
<UDS  service=ReadDataByIdentifierPositiveResponse |<UDS_RDBIPR  dataIdentifier=0x3 |<Raw  load='\\xc0\\xa8\\xb2\x01\\xff\\xff\\xfc\x00' |>>>
```

Now:
```python
>>> pkt = Uds(bytes.fromhex("620003c0a8b201fffffc00"))
>>> print(repr(pkt))
<UdsRdbiPr  service=ReadDataByIdentifierPositiveResponse dataIdentifier=0x3 dataRecord=<Payload3  ip=192.168.178.1 mask=255.255.252.0 |> |>
```

In my opinion, the new implementation is more handy, faster to type and implements the standard in a more accurate way.

Now the big question is, how could this change (once it's ready) be integrated in Scapy. Currently (in development), I use a different class name (Uds... instead of UDS) for the new layer, but to follow the general patterns of Scapy, the long term goal should be to keep the layer name uppercase. 

What do you suggest for getting this new layer implementation into Scapy? 
It will break the existing way of creating UDS packets, which means, some users may need to update their code.

### CODE REVIEW

**Comment by polybassa** [comment]
My plan is, to apply this new structure also to other automotive layers like `GMLAN, OBD, XCP`

File: scapy/fields.py [hunk 1]
```diff
@@ -587,6 +592,62 @@ def fld(self):
         return self._find_fld()
 
 
+class MultiplePacketField(MultipleTypeField):
```
Comment (gpotter2) [inline scapy/fields.py:595]: can you detail what this is for?

Comment (polybassa) [reply scapy/fields.py:595]: This field allows having multiple different packet classes based on conditions. Pretty much like `bind_layers(...)`.
I need a way to dynamically modify the "sub-"protocol of UDS. Basically the first few bytes in a UDS Packet are defined by the standard, but "complex" payloads are always OEM dependent. This means, I need special packets for payloads. In the past, I abused the `bind_layers()` function for this and therefore created all the fake-layers. With this Field, I can dynamically modify the packet class of this field based on certain conditions of the UDS packet.


Here is an example:
```
class Payload3(Packet):
    fields_desc = [
        IPField("ip", 0),
        IPField("mask", 1)
    ]

UdsRdbiPr.bind_data_record(Payload3, 0x3)

pkt = UdsRdbiPr(dataRecord=Payload3(ip="127.0.0.1")
print(repr(pkt))
>>> <UdsRdbiPr  dataIdentifier=0x3 dataRecord=<Payload3  ip=127.0.0.1 |> |>
```

Comment (gpotter2) [reply scapy/fields.py:595]: So I understand what you're trying to do, and this hasn't really been the "Scapy-way" of doing things. Take 802.11 for instance, it's a bunch of `Dot11` layers following each other but it "should" be just one packet.

I don't really like changing the paradigm. I understand what you're doing, but I would think it's confusing to have 2 very different ways of doing one thing within the same tool.

What do you think?

Comment (polybassa) [reply scapy/fields.py:595]: I agree, overall it should be consistent. I didn't know that `Dot11` also has "fake" layers.

I'm ok with the layer as it is, I just thought it's not Scapy like to have this many fake layers in UDS (and other automotive protocols). So I experimented a bit and found this way.
What I do with this field is basically
```
ConditionalField(PacketField(), x=1)
ConditionalField(PacketField(), x=2)
```

Comment (polybassa) [reply scapy/fields.py:595]: So if you think there is no need for a `MultiplePacketField` in Scapy, I will close this PR.

Comment (gpotter2) [reply scapy/fields.py:595]: Typically we already have two ways of doing that:
- A `PacketField` where you pass a function
- A `PacketField` with `dispatch_hook`

I do agree that those aren't obvious, but I don't believe that we need a third way.. so yeah

**Pull request closed without merging**